File: week6/sentimental-readability/readability.py
import re


def main():
    text = input("Enter text: ")
    # Sentences, words, letters.
    s = sentences(text)[:-1]  # Returns list of sentences
    w = words(s)  # Returns number of words
    l = chars(text)  # Returns number of characters

    print(calculate_grade(s, w, l))


def sentences(text) -> list:
    # Split only when theres either !?. * instead of + because otherwise it wont split last sentence and keeps its character.
    sentence_list = re.split('[!?.]\s*', text)
    return sentence_list


def words(sentences) -> int:
    word_count = 0
    for sentence in sentences:
        words_list = re.split('\s', sentence)  # Split list any time theres a whitespace character.
        # print(words_list)
        word_count += len(words_list)

    return word_count


def chars(text) -> int:
    char_list = re.split('[a-zA-Z]', text)
    # print(char_list)
    # print(len(char_list[:-1])) # Length of list containing all the whitespace(not including last punctuation mark because that wont be split)
    return len(char_list)-1


def calculate_grade(s, w, c) -> str:
    # Average number of letters per 100 words.
    avg_let_per_words = (c / w) * 100
    # Average number of sentences per 100 words.
    avg_sen_per_words = (len(s) / w) * 100
    level = 0.0588 * avg_let_per_words - 0.296 * avg_sen_per_words - 15.8

    if round(level) < 1:
        return "Before Grade 1"
    elif (round(level) >= 16):
        return "Grade 16+"

    return f"Grade {round(level)}"


if __name__ == "__main__":
    main()

File: week2/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int grade_level(int letter, int word, int sentence);

int main(void)
{
    string inp = get_string("Enter text: ");
    //printf("Return: %s\n", inp);
    int characters = count_letters(inp);
    int words = count_words(inp);
    int sentences = count_sentences(inp);
    int grade = grade_level(characters, words, sentences);
    //printf("Characters: %i, Words: %i, Sentences: %i\n", characters, words, sentences);
    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
    printf("Grade %i\n", grade);
    }
}

int count_letters(string text)
{
    int let = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isalpha(text[i]))
        {
            let++;
        }
    }
    return let;
}

int count_words(string text)
{
    int words = 1;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == 32 && text[i + 1] != 32)
        {
            words++;
        }
    }
    return words;
}

int count_sentences(string text)
{
    int sentences = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == 33 || text[i] == 46 || text[i] == 63)
        {
            sentences++;
        }
    }
    return sentences;
}

int grade_level(int letter, int word, int sentence)
{
    float average_letter = ((float)letter / (float)word) * 100; //per 100 words
    //printf("Average letters per 100 words: %f\n", average_letter);
    float average_sentence = ((float)sentence / (float)word) * 100; //per 100 words
    //printf("Average sentences per 100 words: %f\n", average_sentence);
    float index = (0.0588 * average_letter - 0.296 * average_sentence - 15.8);
    //printf("Grade: %f\n", index);
    return round(index);
}